### ISSUE TITLE
Input.get_joy_axis reporting at half precision steps

### ISSUE THREAD

**Issue by DJKero** [godotengine/godot#42876]
**Godot version:**
3.2.3.stable.official

**OS/device including version:**
Windows 10 Pro N 1903 18362.1139

**Issue description:**
Input.get_joy_axis method is only updating the value per 2 steps based on 0 to 255 range.
If you see the axis values in a software like XPadder, an axis is represented in 0 to 65535 range and you see the increments of 256, but Godot only updates in multiples of 512.

**Comment by Calinou** [comment]
cc @madmiraal

**Comment by madmiraal** [comment]
I'm struggling to understand the issue. `Input.get_joy_axis()` returns a [float](https://docs.godotengine.org/en/stable/classes/class_input.html#class-input-method-get-joy-axis) between -1 and 1 not a discrete value in multiples of 512.

On Windows this value is extracted using the Windows API's [`XInputGetState`](https://docs.microsoft.com/en-us/windows/win32/api/xinput/nf-xinput-xinputgetstate) function, which uses the [`XINPUT_GAMEPAD`](https://docs.microsoft.com/en-us/windows/win32/api/xinput/ns-xinput-xinput_gamepad) structure. As stated in the [documentation](https://docs.microsoft.com/en-us/windows/win32/api/xinput/ns-xinput-xinput_gamepad), this returns values between 0 and 255 for triggers and -32768 and 32767 for the left and right, x- and y-axes i.e. a range of 65535. Godot stores these range values as constants:
https://github.com/godotengine/godot/blob/9dad483920ffa3c32a03f1f4cc869b3f7d1cbe0e/platform/windows/joypad_windows.h#L69
https://github.com/godotengine/godot/blob/9dad483920ffa3c32a03f1f4cc869b3f7d1cbe0e/platform/windows/joypad_windows.h#L66
It converts the values returned by the OS to a float by dividing the values by these constants:
https://github.com/godotengine/godot/blob/9dad483920ffa3c32a03f1f4cc869b3f7d1cbe0e/platform/windows/joypad_windows.cpp#L457-L473
At no stage does it use multiples of 512.

**Comment by DJKero** [comment]
Let me try to explain myself, sorry if I'm not explaining it well enough by the way.

The issue I'm experiencing is that it seems that the method isn't reflecting the changes in the internal value (-32768 to 32767) when it changes in steps of 256.
For a device with 0-255 precision, 1 of its steps would be a change around 256, however the method doesn't reflect that in the value returned, it only updates when the device moves 2 steps from a correctly detected state.

**Comment by DJKero** [comment]
An example just in case:

When sThumbLX is -256 the method should return -0.0078xx but it returns 0.
When sThumbLX is -512 the method should return -0.015625 and it returns -0.015625.
When sThumbLX is -768 the method should return -0.0234xx but it returns -0.015625.
When sThumbLX is -1024 the method should return -0.03125 and it returns -0.03125.

You get the idea, hah

**Comment by madmiraal** [comment]
@DJKero If what you're suggesting were true, then, when sThumbLX was at the negative limit i.e. -32768, the value Godot would return would be -0.5, but it doesn't, it returns -1; as using any joystick will show.

This leads me to believe that the values of sThumbLX are not what you assert, and that the problem lies with how you're setting the values for sThumbLX.

**Comment by DJKero** [comment]
No, because:
-32768 / MAX_JOY_AXIS equals -1.0
The issue is in between the steps, you cannot test this with a common joystick.

**Comment by DJKero** [comment]
Yeah you can reproduce this even with a cheapo steering wheel and some software that lets you see the value of the X axis like XPadder.

**Comment by DJKero** [comment]
I don't understand what that algorithm does, sorry.
I'll check with other software like XPadder to double check the values.

![2020-10-30_16-00-47](https://user-images.githubusercontent.com/2711997/97746610-64fd4280-1ac9-11eb-875e-4813af24b630.png)
![2020-10-30_16-00-52](https://user-images.githubusercontent.com/2711997/97746619-6890c980-1ac9-11eb-82e8-15903925cd3f.png)
![2020-10-30_16-00-57](https://user-images.githubusercontent.com/2711997/97746621-6af32380-1ac9-11eb-82fb-53c65c0bdeaa.png)
![2020-10-30_16-01-03](https://user-images.githubusercontent.com/2711997/97746626-6c245080-1ac9-11eb-93a6-45fe3fe2b58f.png)
![2020-10-30_16-01-17](https://user-images.githubusercontent.com/2711997/97746630-6d557d80-1ac9-11eb-9083-96d3f3389f41.png)

**Comment by DJKero** [comment]
Here's proof there's something wrong:

![godot windows opt tools 64_2020-10-30_16-12-50](https://user-images.githubusercontent.com/2711997/97747602-ffaa5100-1aca-11eb-8364-656ad76497b3.png)
![2020-10-30_16-13-04](https://user-images.githubusercontent.com/2711997/97747607-00db7e00-1acb-11eb-8eb7-83262e3ebf8c.png)
![2020-10-30_16-13-15](https://user-images.githubusercontent.com/2711997/97747612-020cab00-1acb-11eb-813a-39bd995bbecf.png)
![2020-10-30_16-13-29](https://user-images.githubusercontent.com/2711997/97747615-033dd800-1acb-11eb-81ab-93f71c953299.png)
![2020-10-30_16-13-37](https://user-images.githubusercontent.com/2711997/97747621-05079b80-1acb-11eb-8ffa-8f0b19eb928d.png)
![2020-10-30_16-13-46](https://user-images.githubusercontent.com/2711997/97747625-05a03200-1acb-11eb-9c2d-82c6f28e2759.png)
![2020-10-30_16-14-00](https://user-images.githubusercontent.com/2711997/97747629-06d15f00-1acb-11eb-8c22-08f9195c33ea.png)

**Comment by zyp** [comment]
Hi, I'm the creator of the arcin IO board that @DJKero is using and they asked me to help track down this issue.

The arcin is sending the axis values as an 8-bit number in the range 0..255.

I started out by attempting to reproduce the problem, doing `(get_joy_axis() + 1) / 2 * 255` to scale the -1..1 range back to the original 0..255 range and found it to be moving in steps of three, but sometimes counting 0, 3, 6… and other times counting 1, 4, 7… and so on.

Going through the code, I found the following snippet in `InputDefault::joy_axis()`:
``` cpp
	if (p_value.value > joy.last_axis[p_axis]) {

		if (p_value.value < joy.last_axis[p_axis] + joy.filter) {

			return;
		}
	} else if (p_value.value > joy.last_axis[p_axis] - joy.filter) {

		return;
	}
```
The value here is scaled to a 0..1 range, and `joy.filter` is hardcoded to 0.01. Since a single step is a little bit smaller than 0.004 in this scale, any update event with a smaller difference than three steps from the last reported value is dropped, which is consistent with the behavior I see. Changing `joy.filter` to 0 lets single steps through.

In other words, this is a feature, not a bug, presumably added to get rid of jitter from noisy joysticks. The problem is that it sacrifices precision and the filter threshold is as far as I can see hardcoded in the source, so it can't be changed by users needing the full precision.

**Comment by DJKero** [comment]
I could try to help expose this through the API and learn in the process if you see fit.

**Comment by madmiraal** [comment]
> The value here is scaled to a 0..1 range, and joy.filter is hardcoded to 0.01. Since a single step is a little bit smaller than 0.004 in this scale, any update event with a smaller difference than three steps from the last reported value is dropped, which is consistent with the behavior I see. Changing joy.filter to 0 lets single steps through.

@zyp Yes, I think that's it. Looks like the filter was added on the assumption that the input would be continuous and not in steps; so it's not needed.

BTW Depending on the axis, the value can also be scaled in the -1...1 range with steps of 0.008, but the principal is the same.

**Comment by DJKero** [comment]
Thanks @madmiraal @zyp  @akien-mga 😁
